luigi/rpc.py

```diff
@@ -144,35 +146,36 @@ def __init__(self, url='http://localhost:8082/', connect_timeout=None):
         else:
             self._fetcher = URLLibFetcher()
 
-    def _wait(self):
+    def _get_retry_decorator(self):
+        def retry_logging(retry_state):
+            logger.warning("Failed connecting to remote scheduler %r", self._url, exc_info=True)
+            logger.info("Retrying attempt %r of %r (max)" % (retry_state.attempt_number, self._rpc_retry_attempts))
+
         if self._rpc_log_retries:
-            logger.info("Wait for %d seconds" % self._rpc_retry_wait)
-        time.sleep(self._rpc_retry_wait)
+            return retry(wait=wait_fixed(self._rpc_retry_wait),
+                         stop=stop_after_attempt(self._rpc_retry_attempts),
+                         reraise=True,
+                         after=retry_logging)
+        return retry(wait=wait_fixed(self._rpc_retry_wait),
+                     stop=stop_after_attempt(self._rpc_retry_attempts),
+                     reraise=True)
 
     def _fetch(self, url_suffix, body):
         full_url = _urljoin(self._url, url_suffix)
-        last_exception = None
-        attempt = 0
-        while attempt < self._rpc_retry_attempts:
-            attempt += 1
-            if last_exception:
-                if self._rpc_log_retries:
-                    logger.info("Retrying attempt %r of %r (max)" % (attempt, self._rpc_retry_attempts))
-                self._wait()  # wait for a bit and retry
-            try:
-                response = self._fetcher.fetch(full_url, body, self._connect_timeout)
-                break
-            except self._fetcher.raises as e:
-                last_exception = e
-                if self._rpc_log_retries:
-                    logger.warning("Failed connecting to remote scheduler %r", self._url,
-                                   exc_info=True)
-                continue
-        else:
+        scheduler_retry = self._get_retry_decorator()
+
+        @scheduler_retry
+        def __fetch(full_url, body):
```

Comment:
Can we avoid creating this inner function every time we can `_fetch`? Maybe just make it a normal static method. Also `__` as prefix doesn't seem to be needed: https://en.wikipedia.org/wiki/Name_mangling#Python

Reply:
@honnix I tried to fix this in 5150504 
Which do you prefer original one or fixed?